src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,143 @@
+import time
+import numpy as np
+import matplotlib.pyplot as plt
+
+# Define variables
+API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
+MODELS = "CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S"
+
+
+# Create functions
+def call_API_aux(url): 
+    r = requests.get(url)
+
+    # Check the connections
+    if r.status_code == 200:
+        print("Connection without errors")
+        return r.json()  # correct connection
+
+    else:
+        print(
+            "Number Error: " + str(r.status_code) + "\nType: " + r.text
+        )  # Error explanation
+
+    # Check rate limit error and make cool off(Wait 3 seconds)
+    for _ in range(3):
+        if r.status_code == 429:
+            print(
+                "ERROR: "
+                + str(r.status_code)
+                + " \nType: Max rate limit, you need to wait a bit "
+            )
+            time.sleep(10)
+            r = requests.get(url)
+        continue
+    return None
+
+
+# Obtain data from a city from the API
+def get_data_meteo_api(city, start_date, end_date, temporal_resolution):
+    coordinates = COORDINATES.get(city)
+
+    if coordinates:
+        lat = coordinates["latitude"]
+        long = coordinates["longitude"]
+
+        # New request to the API
+        url = f"{API_URL}latitude={lat}&longitude={long}&start_date={start_date}&end_date={end_date}&models={MODELS}&temporal_resolution={temporal_resolution}&daily={VARIABLES}"
+        return call_API_aux(url)
+    else:
+        print("Coordinate error")
+        return None
+
+
+def compute_average(data):
+    return np.mean(data)
+
+
+def compute_dispersion(data):
+    return np.std(data)
+
+
+# It is not working
+def plot_results(
+    city,
+    temperature_average,
+    temperature_dispersion,
+    precipitation_average,
+    precipitation_dispersion,
+    soil_moisture_average,
+    soil_moisture_dispersion,
+):
+    variables = ["T", "P", "H"]
+    # Create figure and axes
+    fig, gp = plt.subplots()
+    x = range(len(variables))
+
 
 def main():
-    raise NotImplementedError
+    # raise NotImplementedError
+    temporal_resolution = "year"
+    start_date = "1950-01-01"
+    end_date = "1951-12-31"
+    # models = "CMCC_CM2_VHR4"
+
+    # Go through each city in the COORDINATES list and call the function to take each data.
+    for city in COORDINATES.keys():
+        city_data = get_data_meteo_api(city, start_date, end_date, temporal_resolution)
+
+        # print(city_data)
+        print("Dates of: " + city + " " + str(city_data))
+
+        # I have errors in this part of the code because I can´t axcess to this variables (KeyError: 'temperature_2m_mean')
```

Comment:
- Have you gone through the API documentation to understand how does the response look like? Have you explore in python shell how it looks like?

Let's break down the problem:
- We have a KeyError: this means that we have a dictionary which is missing the key we are trying to access
- Therefore, let's check how city_data looks like with city_data.keys(). We observe that there are several keys but none is temperature_2m_mean. Checkig the documentation API or playing around we can see that key `daily` contains the data we are lookig for.
- In addition, since we have several models we have the data of each of those models.

Explore a bit this and read the API doc and rework this part with this information. If you are still lost drop a message over slack